### PR TITLE
Migrate SVM SMO to DPC++ interfaces

### PR DESCRIPTION
**Pull request by vlad-nazarov**

### CODE REVIEW

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+                      const pr::ndarray<Float, 1>& labels,
+                      const std::int64_t row_count,
+                      const std::int64_t ws_count,
+                      const std::int64_t max_inner_iter,
+                      const Float C,
+                      const Float eps,
+                      const Float tau,
+                      pr::ndview<Float, 1>& alpha,
+                      pr::ndview<Float, 1>& delta_alpha,
+                      pr::ndview<Float, 1>& f,
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:119]: you can rename to `grad`

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    const Float* labels_ptr = labels.get_data();
+    const Float* kernel_values_ptr = kernel_values.get_data();
+    const std::uint32_t* ws_indices_ptr = ws_indices.get_data();
+    Float* alpha_ptr = alpha.get_mutable_data();
+    Float* delta_alpha_ptr = delta_alpha.get_mutable_data();
+    Float* f_ptr = f.get_mutable_data();
+    Float* f_diff_ptr = f_diff.get_mutable_data();
+    std::uint32_t* inner_iter_count_ptr = inner_iter_count.get_mutable_data();
+
+    constexpr std::uint32_t max_sg_size = 64;
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:146]: do you want have as max{wg_size / sg_size, sg_size } ?

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+                if (i == 0) {
+                    const Float max_f = local_vars_ptr[0].max_val_ind.value;
+
+                    /* for condition check: m(alpha) >= max_f */
+                    local_vars_ptr[0].local_diff = max_f - ma;
+                    if (inner_iter == 0) {
+                        local_vars_ptr[0].local_eps =
+                            sycl::fmax(eps, local_vars_ptr[0].local_diff * Float(1e-1));
+                        f_diff_ptr[0] = local_vars_ptr[0].local_diff;
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:208]: I here missed a mistake in due time. It is always correct to compute the difference, not only at the first iteration. We fixed  in CPU for 2021.4 release. See [here](https://github.com/oneapi-src/oneDAL/blob/master/cpp/daal/src/algorithms/svm/svm_train_thunder_impl.i#L467)
You need to move diff before if:
```c++
local_vars_ptr[0].local_diff = max_f - ma;
if (inner_iter == 0) {
...
}
item.barrier(sycl::access::fence_space::local_space);
if (local_vars_ptr[0].local_diff < local_vars_ptr[0].local_eps) {
    break;
}

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+struct local_variables {
+    Float delta_b_i;
+    Float delta_b_j;
+    Float local_diff;
+    Float local_eps;
+    enumerate_value<Float> max_val_ind;
+};
+
+template <typename Float>
+inline void reduce_arg_max(sycl::nd_item<1> item,
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:51]: [optional] This functionality can be useful not only here. Maybe think about where you could put in a common place?

**Review by PetrovKP — COMMENTED** [review]
Looks good overall!

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp (hunk 4 above)
Comment (vlad-nazarov) [reply cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:51]: Where else is it used?

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp (hunk 1 above)
Comment (vlad-nazarov) [reply cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:119]: Do you thing name grad is better?

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp (hunk 2 above)
Comment (vlad-nazarov) [reply cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:146]: At submit we don't have information about subgroups, so we can't use this information

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp (hunk 1 above)
Comment (PetrovKP) [reply cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:119]: Yes

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp (hunk 4 above)
Comment (PetrovKP) [reply cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:51]: For common place - maybe primitives... to use in other algorithms

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+                           Float* objective_func,
+                           enumerate_value<Float>* sg_cache,
+                           enumerate_value<Float>& result) {
+    auto sg = item.get_sub_group();
+
+    const std::uint32_t local_id = item.get_local_id(0);
+    const std::uint32_t wg_size = item.get_local_range()[0];
+    const std::uint32_t sg_size = sg.get_local_range()[0];
+    const std::uint32_t sg_count = wg_size / sg_size;
+    const std::uint32_t sg_id = local_id / sg_size;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:60]: Please use `sg.get_group_id()` or `sg.get_group_linear_id()`. According to the specs it should return the id of the subgroup in the work group.

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    const std::uint32_t sg_local_id = sg.get_local_id();
+
+    const std::uint32_t int_max = dal::detail::limits<std::uint32_t>::max();
+
+    Float x = objective_func[local_id];
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:65]: Shouldn't it be
```suggestion
    Float x = objective_func[sg_local_id];
```
?

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    std::uint32_t res_index = sycl::reduce_over_group(sg, res_max == x ? x_index : int_max, minimum<std::uint32_t>());
+
+    if (sg_local_id == 0) {
+        sg_cache[sg_id].value = res_max;
+        sg_cache[sg_id].index = res_index;
+    }
+
+    item.barrier(sycl::access::fence_space::local_space);
+
+    if (sg_id == 0 && sg_local_id < sg_count) {
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:79]: What about just exiting code if `sg_id != 0`? 

Smth like that:
```
if (sg_id != 0 || sg_local_id >= sg_count)
     return;
```

File: cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+    const std::uint32_t* ws_indices_ptr = ws_indices.get_data();
+    Float* alpha_ptr = alpha.get_mutable_data();
+    Float* delta_alpha_ptr = delta_alpha.get_mutable_data();
+    Float* f_ptr = f.get_mutable_data();
+    Float* f_diff_ptr = f_diff.get_mutable_data();
+    std::uint32_t* inner_iter_count_ptr = inner_iter_count.get_mutable_data();
+
+    constexpr std::uint32_t max_sg_size = 64;
+
+    const sycl::nd_range<1> nd_range = dal::backend::make_multiple_nd_range_1d(ws_count, ws_count);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp:147]: Looks strange.

**Pull request closed without merging**